File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/reactiveobject/UaReadObjectAttributeTransaction.java
package org.opcfoundation.webserver.service.transactions.reactiveobject;

import org.eclipse.milo.opcua.stack.core.AttributeId;
import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.DataValue;
import org.eclipse.milo.opcua.stack.core.types.builtin.QualifiedName;
import org.eclipse.milo.opcua.stack.core.types.builtin.StatusCode;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.UInteger;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.eclipse.milo.opcua.stack.core.types.structured.ReadValueId;
import org.opcfoundation.webserver.addressspace.reactiveobject.UaReactiveObjectType;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManagerReactiveObject;
import org.opcfoundation.webserver.addressspace.nodes.UaObject;
import org.opcfoundation.webserver.service.transactions.base.UaReadTransaction;
import org.opcfoundation.webserver.types.common.UaObjectId;
import org.opcfoundation.webserver.types.common.UaObjectIdentifier;
import org.opcfoundation.webserver.service.message.reactiveobject.ReadObjectAttributeRequest;
import org.opcfoundation.webserver.service.message.reactiveobject.ReadObjectAttributeResponse;
import org.opcfoundation.webapi.service.types.ReadContext;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public class UaReadObjectAttributeTransaction extends UaReadTransaction {
    private final UaObjectIdentifier objectId;
    private final NodeManagerReactiveObject nodeManager;

    public UaReadObjectAttributeTransaction(
            ReadContext context,
            List<Integer> handleIds,
            UaObjectIdentifier objectId,
            NodeManagerReactiveObject nodeManager)
    {
        super(context,handleIds);
        this.objectId = objectId;
        this.nodeManager = nodeManager;
    }

    public CompletableFuture<Void> execute()
    {
        // System.out.println("Read object" + objectId);

        try {
            UaReactiveObjectType objectType = nodeManager.findObjectType(objectId);
            if (null == objectType) throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);

            UaObject instanceDeclaration = nodeManager.findInstanceDeclaration(objectId);

            ReadObjectAttributeRequest request = new ReadObjectAttributeRequest(new UaObjectId(
                    objectId.getId(),
                    instanceDeclaration));

            return objectType.onReadObjectAttributes(request).
                    thenAccept(this::setResults).
                    exceptionally(ex -> buildErrorResponse(ex.getCause()));
        } catch (Exception e) {
            buildErrorResponse(e);
        }

        return CompletableFuture.completedFuture(null);
    }

    private void setResults(ReadObjectAttributeResponse response)
    {
        List<ReadValueId> nodesToRead = getRequestedItems();

        for (ReadValueId item: nodesToRead)
        {
            Variant value = Variant.NULL_VALUE;
            StatusCode statusCode = StatusCode.GOOD;

            if (item.getAttributeId().intValue() == AttributeId.DisplayName.id())
            {
                value = new Variant(response.getDisplayName());
            } else if (item.getAttributeId().intValue() == AttributeId.Description.id()) {
                if (response.getDescription().isNull())
                {
                    statusCode = StatusCode.of(StatusCodes.Bad_NodeAttributesInvalid);
                } else {
                    value = new Variant(response.getDescription());
                }
            } else if (item.getAttributeId().intValue() == AttributeId.NodeId.id()) {
                value = new Variant(item.getNodeId());
            } else if (item.getAttributeId().intValue() == AttributeId.BrowseName.id()) {
                value = new Variant(new QualifiedName(0, response.getBrowseName()));
            } else if (item.getAttributeId().intValue() == AttributeId.NodeClass.id()) {
                value = new Variant(NodeClass.Object.getValue());
            } else if (item.getAttributeId().intValue() == AttributeId.WriteMask.id()) {
                value = new Variant(UInteger.valueOf(0));
            } else if (item.getAttributeId().intValue() == AttributeId.UserWriteMask.id()) {
                value = new Variant(UInteger.valueOf(0));
            } else {
                statusCode = StatusCode.of(StatusCodes.Bad_AttributeIdInvalid);
            }

            results.add(new DataValue(
                    value,
                    statusCode,
                    null,
                    null));
        }
    }
}

File: opcua-webservice-node/src/server/service/transactions/reactiveobject/UaReadObjectAttributeTransaction.ts
import { Attributes, NodeClass, StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, UaDataValue, UaError, UaVariant, UaVariantType } from "opcua-webapi-ts";
import { ReadContext } from "../../..";
import { NodeManagerReactiveObject } from "../../../addressspace/nodemanager/NodeManagerReactiveObject";
import { UaObjectId, UaObjectIdentifier } from "../../../types";
import { ReadObjectAttributeRequest, ReadObjectAttributeResponse } from "../../message";
import { UaReadTransaction } from "../base/UaReadTransaction";

export class UaReadObjectAttributeTransaction extends UaReadTransaction {
    private readonly objectId: UaObjectIdentifier;
    private readonly nodeManager: NodeManagerReactiveObject;

    constructor(
        context: ReadContext,
        handleIds: Array<number>,
        objectId: UaObjectIdentifier,
        nodeManager: NodeManagerReactiveObject,
    ) {
        super(context, handleIds);
        this.objectId = objectId;
        this.nodeManager = nodeManager;
    }

    override async execute(): Promise<void> {
        try {
            const objectType = this.nodeManager.findObjectType(this.objectId);
            if (objectType === null) {
                throw new UaError(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
            }

            const instanceDeclaration = this.nodeManager.findInstanceDeclaration(this.objectId);
            const request = new ReadObjectAttributeRequest(
                new UaObjectId(this.objectId.id, instanceDeclaration),
            );

            const response = await objectType.onReadObjectAttributes(request);
            this.setResults(response);
        } catch (error) {
            this.buildErrorResults(error);
        }
    }

    private setResults(response: ReadObjectAttributeResponse): void {
        const nodesToRead = this.getRequestedItems();
        this._results = [];

        for (const item of nodesToRead) {
            let value = UaVariant.null();
            let statusCode = makeUaStatusCode(StatusCodes.Good);

            if (item.attributeId === Attributes.DisplayName) {
                value = UaVariant.localizedText(response.displayName);
            } else if (item.attributeId === Attributes.Description) {
                if (response.description.text.length === 0) {
                    statusCode = makeUaStatusCode(StatusCodes.BadNodeAttributesInvalid);
                } else {
                    value = UaVariant.localizedText(response.description);
                }
            } else if (item.attributeId === Attributes.EventNotifier) {
                value = UaVariant.integer(response.eventNotifier, UaVariantType.Byte);
            } else if (item.attributeId === Attributes.NodeId) {
                value = UaVariant.nodeId(item.nodeId);
            } else if (item.attributeId === Attributes.BrowseName) {
                value = UaVariant.qualifiedName(response.browseName);
            } else if (item.attributeId === Attributes.NodeClass) {
                value = UaVariant.integer(NodeClass.Object, UaVariantType.UInt32);
            } else if (item.attributeId === Attributes.WriteMask || item.attributeId === Attributes.UserWriteMask) {
                value = UaVariant.integer(0, UaVariantType.UInt32);
            } else {
                statusCode = makeUaStatusCode(StatusCodes.BadAttributeIdInvalid);
            }

            this._results.push(new UaDataValue(value, statusCode));
        }
    }
}